### ISSUE TITLE
Install via language package managers

### ISSUE THREAD

**Issue by GeertJohan** [jdx/mise#580]
Many modern applications can be downloaded, compiled and installed through language package managers. For example:

```bash
cargo install ripgrep@13.0.0
```

and

```bash
go install github.com/derailed/k9s@v0.27.4
```

Executing these will install the tool globally, which is not preferred for rtx, but it's fairly doable to have these tools output the compiled binaries into a temp folder, from where rtx can pick them up and manage them just like tools installed via a plugin.

Would it make sense to add support for installing tools through `cargo` and `go`? To make the idea more visual, the cli could look like this:

```bash
rtx use cargo:ripgrep@13.0.0

rtx use go:github.com/derailed/k9s@v0.27.4
```

This has a number of benefits:
- A huge number of  applications become automatically available for installation/version-pinning/updating via rtx, without needing a plugin.
- Installations like these don't suffer from supply-chain attack via plugin scripts; less auditing/trust needed.
- This may work on platforms/architectures for which a tool doesn't have a compiled binary available but can be compiled locally. (Not all tools publish pre-compiled binaries for either linux or macos, arm64, risc-v, etc.)

There are some downsides as well:
- The user needs to have cargo and/or go installed to use this feature. (Ofcourse it would be optional.)
- Downloading and compiling source code is often slower than downloading a precompiled binary, and also uses more storage.
- The official releases may have been built using build-tags / configuration which makes the binary release perferable to a local build.

Because of the downsides, plugins for tools would still serve a purpose. I see this as an addition to the plugins, definitely not as a replacement. Users should be able to switch back and forth without any problems.

Anyway, there are a lot of subtleties to work out before starting on an implementation. Both on the usage side, e.g.: What would .rtx.toml look like? Go uses `v` prefix, do we keep that? etc. As well as on the implementation side; Use globally installed cargo/go or the project-local one? Do we cleanup source code after successful installation?

 But before diving into those I'm just wondering if there is support for a feature like this. Thanks! :)

**Comment by jdx** [comment]
Sorry for the super long delay, I think this idea needed to bake in my head a bit. I like it a lot though. I think it makes sense for things that are CLIs like ripgrep.

> Anyway, there are a lot of subtleties to work out before starting on an implementation. Both on the usage side, e.g.: What would .rtx.toml look like? Go uses v prefix, do we keep that? etc. As well as on the implementation side; Use globally installed cargo/go or the project-local one? Do we cleanup source code after successful installation?

rtx aggressively removes any `v` prefixes in versions for consistency across all tools, that is problematic here though since you're planning for that to map directly to a git tag at least in the case of go packages. I don't know but I don't think I want to remove the stripping of `v` generally even if it might make working with go a bit awkward here. (notably this does mean you can always run `rtx use node@v20` for _any_ tool and it will be identical to not having the `v`.

Now in terms of implementation, there are good things and bad. The good news is there has been work to allow for different version "tags" like `prefix:`, `ref:`, and `path:`. These could just be alternate versions of that. That said, that logic is really in need of a refactor and currently is duplicated in several places.

What also needs to happen are different plugin types. We now do this by having a ["core" plugin type and an "external_plugin" type](https://github.com/jdx/rtx/blob/main/src/plugins). We can fairly easily extend this by having cargo and go which just need to implement the Plugin trait.

What might be tricky here though is managing the cargo/go toolchain. I'm not sure what the best way to handle that would be but there has been some conversation around something similar for elixir which needs to manage erlang in #852.

Final thought, I actually think this might be possible in just a standard asdf plugin (which should work in asdf too) without modifying rtx, we would just need a very slightly different syntax:

```
rtx use cargo@ripgrep@13.0.0
rtx use go@github.com/derailed/k9s@v0.27.4
```

it's just a little weird with the 2 "@" signs.

**Comment by GeertJohan** [comment]
Hi, no worries about the late reply, happy you like the idea :)

I really like your idea of re-using the plugin infrastructure for this. That would allow other 'package/binary managers' to integrate as well. I think rtx does need to be aware of some slight differences between a 'normal' plugin and a 'multi plugin'. (Is that a good name? 'multi plugin'?). `rtx` needs to be aware that `github.com/derailed/k9s@v0.1.0` and `github.com/derailed/k9s@v0.2.0` are two different versions of the SAME binary. Whereas `github.com/foo:v1.0.0` is a different binary from `github.com/bar:v1.0.0`, and could both be active (`rtx use ..`) at the same time. `rtx` should do the bookkeeping there, but the multi-plugin does the downloading/listing.
This does mean that it's not automatically compatible with asdf, but could be relatively easy for them to implement this as well?

CLI syntax wise, what do you think about using a `:` to separate the multi-provider from the package name?
```
rtx use cargo-bin:ripgrep@13.0.0
rtx use go-bin:github.com/derailed/k9s@v0.27.4
```

**Comment by GeertJohan** [comment]
Probably easiest to go with tool dependencies and `bin/list-dependent-tools`, because then rtx doesnt have to know about the binary lookup (i.e. which tool has which binary). And listing binary dependencies doesn't really have a benefit otherwise. We just want to know when to re-install stuff when a depended-on tool verison was bumped.

**Comment by jdx** [comment]
I came across this which does what you're looking for with python: https://github.com/amrox/asdf-pyapp

**Comment by jdx** [comment]
closing in favor of https://github.com/jdx/mise/discussions/1250
